File: 1299.replace-elements-with-greatest-element-on-right-side.java
/*
 * @lc app=leetcode id=1299 lang=java
 *
 * [1299] Replace Elements with Greatest Element on Right Side
 */

// @lc code=start
class Solution {
    public int[] replaceElements(int[] arr) {
        if (arr.length == 1) return new int[] { -1 };
        int max = -1;
        for (int ix = arr.length - 1; ix != -1; --ix) {
            int tmp = arr[ix];
            arr[ix] = max;
            max = Math.max(tmp, max);
        }
        return arr;
    }
}
// @lc code=end

File: 1299.replace-elements-with-greatest-element-on-right-side.cpp
/*
 * @lc app=leetcode id=1299 lang=cpp
 *
 * [1299] Replace Elements with Greatest Element on Right Side
 */

// @lc code=start
#include <vector>

using namespace std;

class Solution {
public:
    vector<int> replaceElements(vector<int>& arr) {
        auto m_max = arr.back(); arr.back() = -1;
        if (arr.size() == 1) return arr;
        for (auto iter = arr.rbegin() + 1; iter != arr.rend(); ++iter) {
            auto tmp = *iter; *iter = m_max; m_max = max(m_max, tmp);
        }
        return arr;
    }
};
// @lc code=end
